### ISSUE TITLE
Nested each_child iterators lose path

### ISSUE THREAD

**Issue by kreynolds** [ohler55/oj#722]
Using the example code snippet, it appears as though while fetching values appears to still work, reading the path does not.

```ruby
    Oj::Doc.open({a:1, c:[2, 3], d:4}.to_json) do |doc|
      doc.each_child('/') do |child|
        puts "#{child.where} -> #{child.fetch}"
        child.each_child do |grandchild|
          puts "#{grandchild.path} -> #{grandchild.fetch}"
        end
      end
    end
```

I expect to see 
```
/a -> 1
/c -> [2, 3]
/c/1 -> 2
/c/2 -> 3
/d -> 4
```
but instead see
```
/a -> 1
/c -> [2, 3]
/c/1 -> 2
/c/2 -> 3
/c/d -> 4
```
Note the final `d` path appears to `fetch` properly but `path` retains `c` when it should not.

**Comment by ohler55** [comment]
I fix and add a test case. Thanks.

**Comment by kreynolds** [comment]
@ohler55 Not to rush, but is there an ETA on this fix? This is currently a blocker for me. I can try fixing it myself if you don't think you'll have time to get to it in the near future

**Comment by ohler55** [comment]
I've been pretty backed up with a move and also work. I'll try to get to it this weekend though now that things are settling down. Sorry for the delay.

**Comment by kreynolds** [comment]
Its no problem at all, I appreciate you taking the time. I'd already started by writing a failing test which I'll put here for your convenience:
```ruby
  def test_nested_each_child
    h = {}
    Oj::Doc.open('{"a":1,"c":[2],"d":3}') do |doc|
      doc.each_child('/') do |child|
        h[child.path] = child.fetch
        child.each_child do |grandchild|
          h[grandchild.path] = grandchild.fetch
        end
      end
    end

    assert_equal({"/a"=>1, "/c"=>[2], "/c/1"=>2, "/d"=>3}, h)
  end
```

**Comment by ohler55** [comment]
Thanks

**Comment by ohler55** [comment]
Just pushed a fix to the "doc-path-loss" branch. If it looks okay I will release.

**Comment by ohler55** [comment]
Released with the fix.

**Comment by kreynolds** [comment]
Yep, this was solved with d3fdc91014b7e1c870f799e717c089b266969307 .. thanks!
